src/pages/node-express/lesson2-a1.mdx

```diff
@@ -0,0 +1,304 @@
+    writer. Finally, write a line at the bottom of the file that calls the
+    `readWrite` function. Test your code by running The file that your code is
+    creating should not be sent to Github, so you should add its filename as
+    another line to your <code>.gitignore.</code>
+  </p>
+</div>
+
+2.  Write another program called `writeWithPromisesThen.js`. You start it the same way, but this time, you use the `.then` style of asynchronous programming. You don’t need to create any functions. Instead you just use cascading .then statements in your mainline, like this:
+
+```js
+ writeFile(...) // write line 1
+ .then(() => {
+    return writeFile(...)  // write line 2.
+    // Return the promise so you can chain the .then statements
+ })
+ .then // write the third line, and follow that with two more .then blocks,
+ // one to call readFile to read it back out, and one to log the data to the screen.
+ ...
+ .catch((error) => {
+     console.log("An error occurred: ", error)
+ })
+```
+
+<p class="ml-8">
+  Test your code by running <code>node writeWithPromisesThen.js</code>. You may
+  want to sprinkle console.log statements in your code so that you understand
+  the order of execution.
+</p>
+
+3.  We want to understand event emitters. First, modify `prompter.js`, to add the following lines above the listen statement:
+
+```js
+server.on("request", (req) => {
+  console.log("event received: ", req.method, req.url);
+});
+```
+
+<p class="ml-8">
+  Then test this (<code>npm run dev</code>) and try with your browser to see the
+  events the server is emitting.
+</p>
+
+4.  Write a program named `customEmitter.js`. In it, create one or several emitters. Then use the emitter `.on` function to handle the events you will emit, logging the parameters to the screen. Then use the emitter `.emit` function to emit several events, with one or several parameters, and make sure that the events are logged by your event handlers. This is your chance to be creative! You could have an event handler that emits a different event to be picked up by a different handler, for example. Here’s a couple tricks to try. You can trigger events with a timer, as follows:
+
+```js
+const EventEmitter = require("events");
+
+const emitter = new EventEmitter();
+
+setInterval(() => {
+  emitter.emit("timer", "hi there");
+}, 2000);
+
+emitter.on("timer", (msg) => console.log(msg));
+```
+
+<p class="ml-8">Or, you could make an async function that waits on an event:</p>
+
+```js
+const EventEmitter = require("events");
+
+const emitter = new EventEmitter();
+
+const waitForEvent = () => {
+  return new Promise((resolve) => {
+    emitter.on("happens", (msg) => resolve(msg));
+  });
+};
+
+const doWait = async () => {
+  const msg = await waitForEvent();
+  console.log("We got an event! Here it is: ", msg);
+};
+
+doWait();
+emitter.emit("happens", "Hello World!");
+```
+
+<p class="ml-8">
+  Don't worry if that last one looks a bit complicated. That's expected as we
```

Comment:
Just a mild bit of rephrasing to subtly make this more postively-toned / growth mindset aligned (instead of 

> I guess that last one is not so easy

)

YMMV though of course!